src/display/svg.js

```diff
@@ -982,8 +982,54 @@ SVGGraphics = class SVGGraphics {
    */
   _makeColorN_Pattern(args) {
     if (args[0] === 'TilingPattern') {
-      warn('Unimplemented pattern TilingPattern');
-      return null;
+      const color = args[1];
+      const operatorList = args[2];
+      const matrix = args[3] || IDENTITY_MATRIX;
+      const [x0, y0, x1, y1] = args[4];
+      const xstep = args[5];
+      const ystep = args[6];
+      const paintType = args[7];
+
+      const tilingId = `shading${shadingCount++}`;
+      const [tx0, ty0] = Util.applyTransform([x0, y0], matrix);
+      const [tx1, ty1] = Util.applyTransform([x1, y1], matrix);
+      const [xscale, yscale] = Util.singularValueDecompose2dScale(matrix);
+      const txstep = xstep * xscale;
+      const tystep = ystep * yscale;
+
+      const tiling = this.svgFactory.createElement('svg:pattern');
+      tiling.setAttributeNS(null, 'id', tilingId);
+      tiling.setAttributeNS(null, 'patternUnits', 'userSpaceOnUse');
+      tiling.setAttributeNS(null, 'width', txstep);
+      tiling.setAttributeNS(null, 'height', tystep);
+      tiling.setAttributeNS(null, 'x', `${tx0}`);
+      tiling.setAttributeNS(null, 'y', `${ty0}`);
+
+      // Save current state.
+      const svg = this.svg;
+      const transformMatrix = this.transformMatrix;
+      const fillColor = this.current.fillColor;
+      const strokeColor = this.current.strokeColor;
+
+      const bbox = this.svgFactory.create(tx1 - tx0, ty1 - ty0);
+      this.svg = bbox;
+      this.transformMatrix = matrix;
+      if (paintType === 2) {
+        const cssColor = Util.makeCssRgb(...color);
+        this.current.fillColor = cssColor;
+        this.current.strokeColor = cssColor;
+      }
```

Comment:
Let's change this to:

```javascript
} else {
  warn(`Unimplemented tiling type ${paintType}`);
}
```

This way we still get notified in the console if an unsupported tiling type is used, which helps for debugging issues.

Reply:
paintType can be only `1` or `2` in the [PDF specification](https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdf_reference_archive/pdf_reference_1-7.pdf#page=292).
This implementation also includes processing when `paintType === 1`, so its `else` clause seems unnecessary.

Reply:
You're right. Thank you for clarifying this.